packages/protocol/contracts/common/Signatures.sol

```diff
@@ -49,4 +49,44 @@ library Signatures {
     bytes32 prefixedHash = ECDSA.toEthSignedMessageHash(messageHash);
     return ECDSA.recover(prefixedHash, signature);
   }
+
+  /**
+  * @notice Given a domain separator and a structHash, construct the typed data hash
+  * @param eip712DomainSeparator Context specific domain separator
+  * @param structHash hash of the typed data struct
+  * @return The EIP712 typed data hash
+  */
+  function toEthSignedTypedDataHash(bytes32 eip712DomainSeparator, bytes32 structHash)
```

Comment:
I don't know off the top of my head, but would it be possible for calllers to pass something closer to the struct itself instead of the hash? Ideally we'd mirror the sign interface? The encoding + hashing is ideally something a caller shouldn't have to think about?

Reply:
Yeah I tried going in that direction that's why I setup up this function to move more of that hash construction away from the consumer of `Signatrues.sol` but I don't see a way to move away unless we create another level of abstraction for each type of signed data that we want, and introduce a new library contract that's responsible for constructing these hashes, which would mean splitting that functionality away from the `MetaTransactionWallet` contract into somewhere else. That way other contracts can rely on that library to construct the hashes and so on without consuming the full `MetaTransactionWallet` but I don't see a really clear benefit to that.

Reply:
Gotcha, what about just passing the bytes of the `abi.encode`? Or would that not really "hide" the implementation details enough?

Reply:
Doesn't feel like a huge win but would just maybe add some bloat to the traces (not 100% sure how library calls are encoded). 

But I could be convinced. Any other folk want to chime in: @asaj / @m-chrzan ?

Reply:
Yeah, I don't think its a huge difference, and i think its fine to keep this way

Reply:
to me it seems appropriate that the `caller`/`client` would encode and hash the input  off-chain. it's cheaper, more efficient, and a good client would hopefully make it intuitive for the non-technical end-user.

Reply:
merp forgot about email notifications...but deleted my last comment bc it didn't really apply